causal-analysis/TowardsCausalThinking.ipynb

```diff
@@ -0,0 +1,2117 @@
+    "\n",
+    "\n",
+    "Notice - all of these are excellent symptoms. It does not explain why these happened.\n",
+    "\n",
+    "__storegw-health-indicator is a vector__\n",
+    "\n",
+    "$\\mathbf{storegw-health-indicator} = \\begin{bmatrix} ThanosStoreGrpcErrorRate \\\\\\ ThanosStoreSeriesGateLatencyHigh \\\\\\ ThanosStoreBucketHighOperationFailures \\\\\\ ThanosStoreObjstoreOperationLatencyHigh   \\end{bmatrix}$\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "7b9b4934",
+   "metadata": {},
+   "source": [
+    "#### Analysis with Causal Model (aka metrics)\n",
+    "\n",
+    "In this case, we will assign metrics to each of the nodes - and will model relationships between the metrics as mentioned earlier. Based on this, the above causal graphs can be slightly change."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "597034c6",
+   "metadata": {},
+   "source": [
+    "##### Compactor\n",
+    "\n",
+    "A causal model looks like - \n",
+    "\n",
+    "<!--\n",
+    "$compactor-health = \\mathbf{f}(is-compactor-out-ofspace,is-there-a-bad-block)$\n",
+    "-->\n",
+    "$\\mathbf{compactor-health} = \\mathbf{f}(is-compactor-out-ofspace,is-there-a-bad-block)$\n",
+    "\n",
+    "How do we capture if there is a bad block - this may be hard and we may require some approximations perhaps?\n",
+    "A bad compactor is a train wreck waiting to happen. This can quickly snowball into lot more issues if not taken care off right away.\n",
+    "  "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "18e19d13",
+   "metadata": {},
+   "source": [
+    "##### Receiver\n",
+    "\n",
+    "\n",
+    "A causal model looks like - \n",
+    "\n",
+    "$\\mathbf{receiver-health} = \\mathbf{f}(time-series, number-of-receiver-replicas, is-receiver-pv-out-of-space, can-receiver-reach-obj-store)$\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "cebd6387",
+   "metadata": {},
+   "source": [
+    "##### Querier\n",
+    "\n",
+    "\n",
+    "A causal model looks like - \n",
+    "\n",
+    "$\\mathbf{querier-health} = \\mathbf{f}(number-of-time-series-per-query, number-of-simaltaneous-queries, number-of-querier-replicas, receiver-health, query-cache-health)$\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "6fdf4714",
+   "metadata": {},
+   "source": [
+    "##### Ruler\n",
+    "\n",
+    "A causal model looks like - \n",
+    "\n",
+    "$\\mathbf{ruler-health} = \\mathbf{f}(number-of-rules, number-of-timeseries, number-of-simaltaneous-queries, number-of-querie-replicas,querier-health)$\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c7e03b3c",
+   "metadata": {},
+   "source": [
+    "##### Store Gateway\n",
+    "\n",
+    "\n",
+    "\n",
+    "A causal model looks like - \n",
+    "\n",
+    "$\\mathbf{storegw-health} = \\mathbf{f}(history-configuration, number-of-timeseries, number-of-simaltaneous-queries,store-gw-cache-health)$"
```

Comment:
Hmm not sure what history configuration is here, maybe retention?
Should include number of total blocks, and number of storegateway shards. Also PV size would need to be included as that is where it builds up block index cache
```suggestion
    "$\\mathbf{storegw-health} = \\mathbf{f}(history-configuration, number-of-blocks, number-of-storegw-shards, number-of-timeseries-per-storeapi-call, number-of-simultaneous-queries,store-gw-cache-health, storegw-space-on-pv)$"
```